### PR TITLE
Encouraging React class usage convention

### PR DESCRIPTION
**Pull request by ziahamza**
Closes this discussion https://github.com/rainforestapp/frontend-conventions/issues/19

### CODE REVIEW

File: ReactComponents.md [hunk 1]
```diff
@@ -45,18 +43,50 @@ const Bouncer = ({age}) => {
...
 ## 3 - Component classes
 
-There are a bunch of reasons why you might need a compon
+Use component classes whenever local state and event handlers are needed. ES7 reduces a lot of boiler plate to use React Classes compared to recompose or other alternatives. 
+
+A simple example of the usage would be:
+```
+class Button extends React.Component {
+  static defaultProps = { initialCount: 0 }; 
+  static state = { count: this.props.initialCount };
```
Comment (jkaipr) [inline ReactComponents.md]: I think the `static` shouldn't be here.

File: ReactComponents.md [hunk 2]
```diff
@@ -45,18 +43,50 @@ const Bouncer = ({age}) => {
...
+
+## 4 - HOCs
 
-... to be written
+Use HOCs when React Classes / Stateless components are not enough or too restrictive to implement certain behaviours. A good set usecases would be
 
+- Making sure state is persistant across user sesstion. See [WithPersistentState](https://github.com/rainforestapp/regenwald/tree/develop/src/app/utilities/withPersistentState)
+- Subscribing / transforming data from redux. See [connect](http://redux.js.org/docs/basics/UsageWithReact.html)
+- Subscribing to action state changes. See [withButtonState](https://github.com/rainforestapp/regenwald/tree/develop/src/app/components/Button/withButtonState.js)
+- Subscribing
```
Comment (jkaipr) [inline ReactComponents.md]: Can be removed or maybe point to `withRouter`.

Comment (ziahamza) [reply ReactComponents.md]: yea, I knew I was missing something!

File: ReactComponents.md [hunk 3]
```diff
@@ -45,18 +43,50 @@ const Bouncer = ({age}) => {
...
+  }
+}
+```
+
+## 4 - HOCs
 
-... to be written
+Use HOCs when React Classes / Stateless components are not enough or too restrictive to implement certain behaviours. A good set usecases would be
 
+- Making sure state is persistant across user sesstion. See [WithPersistentState](https://github.com/rainforestapp/regenwald/tree/develop/src/app/utilities/withPersistentState)
```
Comment (pscohn) [inline ReactComponents.md]: `sesstion`

Comment (pscohn) [reply ReactComponents.md]: `persistent`

**Comment by pscohn** [comment]
Maybe a good rule of thumb is to only ever write/import reusable hocs, never write them in your component?

File: ReactComponents.md [hunk 4]
```diff
@@ -45,18 +43,50 @@ const Bouncer = ({age}) => {
...
 ```
 
 ## 3 - Component classes
 
-There are a bunch of reasons why you might need a compon
+Use component classes whenever local state and event handlers are needed. ES7 reduces a lot of boiler plate to use React Classes compared to recompose or other alternatives. 
+
+A simple example of the usage would be:
+```
+class Button extends React.Component {
```
Comment (jkaipr) [inline ReactComponents.md]: One more nit, going to rename this to `Counter`.

**Comment by ziahamza** [comment]
I like that idea @pscohn . so only use HOCs but dont build new abstractions with them!

**Comment by ziahamza** [comment]
@pscohn added this change in this PR: https://github.com/rainforestapp/frontend-conventions/pull/26

**Comment by pscohn** [comment]
@ziahamza 👍  yeah I would say there are probably cases where it makes sense to extract logic out and make a new hoc, but generally don't define them in component files - if only one component needs it it probably doesn't need the abstraction

**Pull request merged**